ingress/loadbalancer.go

```diff
@@ -17,19 +37,181 @@ type LoadBalancerEntry struct {
+	cmd       *exec.Cmd
+	signaller Signaller
 }
 
-func (lb *noopLoadBalancer) Update(entries []LoadBalancerEntry) error {
-	log.Debugf("Updating loadbalancer for %v", entries)
-	return nil
+// Used for generating nginx cofnig
+type loadBalancerTemplate struct {
+	Port       int
+	Entries    []LoadBalancerEntry
+	WorkingDir string
+}
+
+// DefaultSignaller sends os signals
+type DefaultSignaller struct {
+}
+
+// Sigquit sends a SIGQUIT to the process
+func (s *DefaultSignaller) Sigquit(p *os.Process) error {
+	return p.Signal(syscall.SIGQUIT)
+}
+
+// Sighup sends a SIGHUP to the process
+func (s *DefaultSignaller) Sighup(p *os.Process) error {
+	return p.Signal(syscall.SIGHUP)
 }
 
 // NewLB creates a new LoadBalancer
-func NewLB() LoadBalancer {
-	return &noopLoadBalancer{}
+func NewLB(nginxLocation, nginxConfDir string, port int, signaller Signaller) LoadBalancer {
+	log.Infof("Creating nginx with %s binary in %s", nginxLocation, nginxConfDir)
+	return &nginxLoadBalancer{
+		binary:    nginxLocation,
+		confDir:   nginxConfDir,
+		port:      port,
+		signaller: signaller}
+}
+
+func (lb *nginxLoadBalancer) Start() error {
+	// Write out initial configuration
+	lb.update(LoadBalancerUpdate{Entries: []LoadBalancerEntry{}})
+
+	configLocation := lb.confDir + "/nginx.conf"
+	log.Infof("Using config file %s", configLocation)
+	cmd := exec.Command(lb.binary, "-c", configLocation)
+
+	cmd.Stdout = os.Stdout
+	cmd.Stdin = os.Stdin
+
+	lb.cmd = cmd
+	err := cmd.Start()
+	if err != nil {
+		log.Error("Unable to start nginx", err, cmd)
+		return err
+	}
+
+	go func() {
+		log.Info("Nginx has exited: ", cmd.Wait())
+	}()
+
+	log.Infof("Nginx pid %s", cmd.Process.Pid)
+	return nil
+}
+
+func (lb *nginxLoadBalancer) Stop() error {
+	log.Info("Shutting down process %s", lb.cmd.Process.Pid)
+	err := lb.signaller.Sigquit(lb.cmd.Process)
+	if err != nil {
+		log.Error("Error shutting down nginx", err)
+		return err
+	}
+	return nil
+}
+func (lb *nginxLoadBalancer) Update(entries LoadBalancerUpdate) (bool, error) {
+	updated, err := lb.update(entries)
+	if err != nil {
+		return updated, err
+	}
+	if updated {
+		err = lb.signaller.Sighup(lb.cmd.Process)
+		if err != nil {
+			log.Warn("Configuration was wirtten but unable to signal nginx to reload configuration", err)
+			return false, err
+		}
+	}
+	return updated, err
+}
+
+func (lb *nginxLoadBalancer) update(entries LoadBalancerUpdate) (bool, error) {
+	log.Debugf("Updating loadbalancer %s", entries)
+	file, err := lb.createConfig(entries)
+	if err != nil {
+		return false, err
+	}
+	existing, err := ioutil.ReadFile(lb.confDir + "/nginx.conf")
+	if err != nil {
+		log.Info("Unable to read existing nginx config file. Assuming it needs creating for the first time.", err)
+		return writeFile(lb.confDir+"/nginx.conf", file)
+
+	}
+	diffOutput, err := diff(existing, file)
+	if err != nil {
+		log.Warn("Unable to diff nginx files", err)
+		return false, err
+	}
+	log.Infof("Diff output %s", string(diffOutput))
```

Comment:
Is there a point to logging the diff if it's empty? Can move this until after the next conditional.